### PR TITLE
fix: pre-compile regex for parsing kafka version

### LINKED ISSUE #2662: Compiling the regular expression every time it is consumed is redundant.
#### Description
The target code:
https://github.com/IBM/sarama/blob/44f6db53dd4590c90f93ec6cafd20763211db52e/utils.go#L302
The function regexp.MustCompile(pattern) will comiple the specified pattern when consuming every time. It is not CPU friendly.

The following is the part of CPU profile(Flame Graph) in production environment. It accounts for 7.6% in CPU usage and should be optimized.
cpu profile

### PR DESCRIPTION
**Pull request by qshuai**
Fix compiling regular pattern redundantly to optimize CPU usage.

Fixes #2662

### CODE REVIEW

File: utils.go [hunk 1]
```diff
@@ -280,6 +280,11 @@ var (
 	}
 )
 
+var (
+	validKafkaVersion1 = regexp.MustCompile(`^0\.\d+\.\d+\.\d+$`)
```
Comment (Gorgonx7) [inline utils.go:284]: Thanks for raising this, I think we could probably be a bit more clear with the variable names here and probably should just add a comment
```suggestion
        //This regex validates that a string complies with the pre kafka 1.0.0 format for version strings, for example 0.11.0.3
	validPreKafka1Version = regexp.MustCompile(`^0\.\d+\.\d+\.\d+$`)
```

File: utils.go [hunk 2]
```diff
@@ -280,6 +280,11 @@ var (
...
+	validKafkaVersion2 = regexp.MustCompile(`^\d+\.\d+\.\d+$`)
```
Comment (Gorgonx7) [inline utils.go:285]:
```suggestion
        //This regex validates that a string complies with the post Kafka 1.0.0 format, for example 1.0.0
	validPostKafka1Version = regexp.MustCompile(`^\d+\.\d+\.\d+$`)
```

File: utils.go [hunk 3]
```diff
@@ -288,18 +293,18 @@ func ParseKafkaVersion(s string) (KafkaVersion, error) {
 	var major, minor, veryMinor, patch uint
 	var err error
 	if s[0] == '0' {
-		err = scanKafkaVersion(s, `^0\.\d+\.\d+\.\d+$`, "0.%d.%d.%d", [3]*uint{&minor, &veryMinor, &patch})
+		err = scanKafkaVersion(s, validKafkaVersion1, "0.%d.%d.%d", [3]*uint{&minor, &veryMinor, &patch})
```
Comment (Gorgonx7) [inline utils.go:296]:
```suggestion
		err = scanKafkaVersion(s, validPreKafka1Version, "0.%d.%d.%d", [3]*uint{&minor, &veryMinor, &patch})
```

File: utils.go [hunk 4]
```diff
@@ -288,18 +293,18 @@ func ParseKafkaVersion(s string) (KafkaVersion, error) {
...
 	} else {
-		err = scanKafkaVersion(s, `^\d+\.\d+\.\d+$`, "%d.%d.%d", [3]*uint{&major, &minor, &veryMinor})
+		err = scanKafkaVersion(s, validKafkaVersion2, "%d.%d.%d", [3]*uint{&major, &minor, &veryMinor})
```
Comment (Gorgonx7) [inline utils.go:298]:
```suggestion
		err = scanKafkaVersion(s, validPostKafka1Version, "%d.%d.%d", [3]*uint{&major, &minor, &veryMinor})
```

**Review by Gorgonx7 — CHANGES_REQUESTED** [review]

**Comment by qshuai** [comment]
@Gorgonx7 Thanks! I have push a new commit fixing these issues

**Review by dnwe — APPROVED** [review]
Thanks for submitting this, changes look good to me

**Pull request merged**